### PR TITLE
issue4399 Extension for running Narayana LRA participants

### PR DESCRIPTION
**Pull request by mmusgrov**
…cipants

#4399 

This PR adds support for an implementation of the eclipse microprofile-lra project which is current RC1 but the project is close to have its first release.

### CODE REVIEW

**Comment by kenfinnigan** [comment]
@gsmet do we have a designated person for reviewing PRs around transactions?

File: integration-tests/narayana-lra/pom.xml [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        </dependency>
+        <dependency>
+            <groupId>io.rest-assured</groupId>
+            <artifactId>rest-assured</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-junit5</artifactId>
+            <version>999-SNAPSHOT</version>
```
Comment (jmartisk) [inline integration-tests/narayana-lra/pom.xml]: I suppose the version shouldn't be here

**Comment by gsmet** [comment]
@kenfinnigan not really. TBH, I'm prioritizing things right now and I don't think we should review new extensions, and probably not until 1.3.0 is out of the door, that's why I haven't looked at that one.

This one is pretty niche IMHO so I would say it can wait a bit (unless there are good reasons to fast track it?).

**Comment by kenfinnigan** [comment]
Was it agreed that new extensions wouldn't be added prior to 1.3?

**Comment by mmusgrov** [comment]
> This one is pretty niche IMHO so I would say it can wait a bit (unless there are good reasons to fast track it?).

I agree that we can wait if it helps get 1.3 out, especially since we are only using a candidate release. But, for the record, I think it is worthwhile me stating that the LRA spec means that services can achieve guaranteed consistency in the presence of failures since outcomes are driven to completion when the system recovers. This feature provides business logic authors with a powerful mechanism for writing robust systems with predictable behaviour. The nearest thing to it in the current suite of specs is MP-FT. I would argue that LRA is as useful as these other non niche specs.

**Comment by mmusgrov** [comment]
> @kenfinnigan not really. TBH, I'm prioritizing things right now and I don't think we should review new extensions, and probably not until 1.3.0 is out of the door, that's why I haven't looked at that one.
> 
> This one is pretty niche IMHO so I would say it can wait a bit (unless there are good reasons to fast track it?).

Now that Quarkus 1.3.0.Final is available would it be possible to get this one reviewed?

**Comment by mmusgrov** [comment]
@gsmet and @kenfinnigan Is there anyone we could ping to take a look at this for us.

**Comment by kenfinnigan** [comment]
It needs a rebase because of changes to the poms.

Once that's done we can get someone to take a look

**Comment by mmusgrov** [comment]
> It needs a rebase because of changes to the poms.
> 
> Once that's done we can get someone to take a look

Thanks Ken

**Comment by mmusgrov** [comment]
I skimmed two of the failing checks and they seemed to be in other extensions. I could not find any way to take a copy of the output of check in order to verify that the narayana-lra extension was okay.

Is it possible to grab a copy of the check output (such as "Quarkus CI / JDK 8 JVM Tests") or to search it online.

**Comment by kenfinnigan** [comment]
Saw an issue that Naryana test got a 503 returned.

Wondering if CI is hitting time limits and things are shutting down?

Any ideas @gsmet?

**Review by xstefank — CHANGES_REQUESTED** [review]
From a functional perspective which I verified this PR by using this extension locally a few insights:

LRA works as expected for JAX-RS cases. I verified most of the spec supported configurations including nested LRAs.

There are two big concerns/issues I can see. The first one is a lack of support for non-JAR-RS LRA participants (spec required in TCK). But this can be done in the later revision.

The second one is that lra-coordinator is always combined with the LRA client in this extension. I don't think this is a good design. There will as many coordinators in the microservices deployment as participants. Users must be careful to configure all microservices to reach the same coordinator otherwise there will be wrong results. But the biggest problem is that the lra-coordinator is packaging a lot of processing (JAX-RS resources, arjuna core) which are for most of the LRA use cases (participants) not required. But because the LRA coordinator exposes the JAX-RS interface it will be always loaded. As this goes in the opposite direction than Quarkus intends (small runtime, only add what is needed) I would suggest having to separate extensions one for the coordinator and one for participants, so users have the option to combine them in one service but are not required to do so.

**Comment by mmusgrov** [comment]
> There are two big concerns/issues I can see. The first one is a lack of support for non-JAR-RS LRA participants (spec required in TCK). But this can be done in the later revision.

Correct, but let's get something out there first so that users can get used to the functionality. If they find they would like to have non-JAX-RS participants later then they can have that in a later revision.

> The second one is that lra-coordinator is always combined with the LRA client in this extension. I don't think this is a good design. There will as many coordinators in the microservices deployment as participants.

That was by design, every participant comes packaged with a coordinator so that it does have to go searching for one.

> Users must be careful to configure all microservices to reach the same coordinator otherwise there will be wrong results.

How did you reach that conclusion, LRA doesn't work that way: the coordinator location/endpoint is in the LRA URI that gets passed around during the business interaction (we use a JAX-RS header for this context). The coordinator for a particular LRA is located at the microservice that created it (unless you configure an external one, which you can do if you want that). The coordinator is available as long as the service initiator is available.

> But the biggest problem is that the lra-coordinator is packaging a lot of processing (JAX-RS resources, arjuna core) which are for most of the LRA use cases (participants) not required.

There is no problem here:

- Firstly there is no processing if the LRA is created by an external coordinator (another microservice).
- Secondly, the processing overhead of managing the lifecycle of an LRA is load balanced around the system for free.
- Thirdly, arjuna is a lean piece of code engineered for safety and performance. Sizewise it only adds in another 70K. And a coordinator is comprised of two JAX-RS endpoints (and JAX-RS is prety lightweight these days), besides JAX-RS is needed by the microservice anyway so it is already bundled with the service. Just because arjunacore is very functional it does not mean it adds bulk.
- [Fourthly, having arjunacore available also gives the participant access to an object store for persisting state. I would imagine that is going to come in quite useful in version 1.x of the LRA spec.]

> But because the LRA coordinator exposes the JAX-RS interface it will be always loaded.

With this approach:
 
- every service can start LRAs without having to be preconfigured with the location of an external coordinator (with the management overhead that would bring). That is a significant win.
- and if a service is using LRA it is almost certain that it will want to create LRAs.

> As this goes in the opposite direction than Quarkus intends (small runtime, only add what is needed) I would suggest having to separate extensions one for the coordinator and one for participants, so users have the option to combine them in one service but are not required to do so.

ArjunaCore only adds less that 70K and for that overhead the user gets a lot of functionality. Don't equate arjuna core with bulk. I did start out with separate extensions but soon realised that it was just not necessary and combining them brings more benefits that drawbacks (if the user wants a dedicated external coordinator he just starts one running and configures a microservice to use it).

**Comment by zhfeng** [comment]
> every service can start LRAs without having to be preconfigured with the location of an external coordinator (with the management overhead that would bring). That is a significant win.

Well, I agree with @mmusgrov about this and it could be helpful to use in the camel-lra which we don't need to launch the external coordinator server. I just wonder it is possible to introduce such option ```quarkus.lra.embed.enabled``` which the default value is ```true``` and could be set to ```false``` if the user want to use the external coordinator and do not expose the LRA endpoints on the localhost. Since the quarkus supports the command mode now and the resteasy could be optional.

**Comment by mmusgrov** [comment]
I reopened issue #4400 to address the reviewers requirements. The current PR satisfies the requirements of the issue, namely to bundle the participant and coordinator functionality in the same extension and therefore it should be reviewed on that basis.

I agree with the reviewer that some users may not want both behaviours in the same service and I have describe the pros and cons of each extension (participant only versus participant + coordinator) in the guide I included in the other PR (#9108 ) but I will paste the reasons here:

== Recap of the LRA Model

Long Running Actions (LRAs) are a relaxed transaction model for loosely coupled services to coordinate work. Different services perform work in the context of an LRA and will confirm or compensate for any actions when the LRA is later finished (we say closed or cancelled). A coordinator manages the lifecycle and membership of LRAs and guarantees that services will be notified when the LRA finishes; even in the presence of network outages or service issues (such as a crashed or hung JVM) by retrying pending notifications until all services have responded, only then will an LRA be deemed finished. Each LRA has its own coordinator although a single coordinator may manage many LRAs. Coordinators expose two JAX-RS endpoints.

After creating an LRA (via the `@LRA` annotation), a JAX-RS service can join with the LRA (via the `@LRA` annotation). Joining an LRA means that the service will be notified when the LRA finishes. Some service will eventually end the LRA (via the `@LRA` annotation).

The lifecycle of an LRA is managed by a (LRA) coordinator which can be in-VM or remote. Services communicate with the coordinator via a JAX-RS interceptor that inspects the annotations on business methods.

The `narayana-lra-participant` only bundles the JAX-RS interceptor with the service. The `narayana-lra extension` also includes coordinator behaviour).

== Pros and Cons of the Two Extensions

The factors to bear in mind when choosing which extension to use include (at least) the following:

1. You do not want the management overhead of having to configure the address of an external coordinator.
2. You do not want the 60K overhead that an embedded coordinator adds to a service.
3. You do not want the overhead of an extra process in the external coordinator case.
4. You want to centrally manage the storage (filesystem or JDBC accessible database) for LRAs (LRAs must be durable in order to survive crashes).
5. You want to centrally manage coordinators.
6. You want to shut down the service when it has finished its work in the context of an LRA (a coordinator has to be available to all services involved in an LRA in order to finish the LRA so if it's embedded with the service it must continue running (or be restarted) until all participants have been notified.
7. You want the LRA initiator (the service that triggered the creation of the LRA) to manage the coordination internally to an organisation.
8. You want the coordinator to have the same security config as your service.

We recommend the `narayana-lra extension` (embedded coordinator) if any of 1, 3, 5, 7, 8 apply.
We recommend the `narayana-lra-participant extension` (external coordinator) if any of 2, 4, 5, 6 apply.

But for ease of use, to quickly get started using LRA we advise that you start out with an embedded coordinator and only consider an external one if any of the above mentioned factors come into effect.

**Comment by nicolaferraro** [comment]
Personally I like the option to have an embedded coordinator in the service that can be optionally used without requiring an external coordinator. It comes handy, especially during development, and it is also a valid deployment option for some use cases.

We had a simplified in-process coordinator in Apache Camel and it has been used extensively e.g. in examples and workshops so far, because of its simplicity.

I'd be happy to rebase on this for the embedded coordinator in Camel-Quarkus, also to have the additional features of a real coordinator. As far as I understand, non-jax-rs APIs will come in another PR later. Looking forward for having them.

**Comment by xstefank** [comment]
So I was convinced that we will start with embedded coordinator + participant extension to get user feedback. We can always split them into two extensions later if needed. After the remaining comments will be addressed I will approve this PR.

File: integration-tests/narayana-lra/pom.xml (hunk 1 above)
Comment (mmusgrov) [reply integration-tests/narayana-lra/pom.xml]: +1

File: extensions/narayana-lra/runtime/src/main/java/io/quarkus/narayana/lra/runtime/NarayanaLRAProducers.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+import javax.enterprise.context.Dependent;
+import javax.enterprise.inject.Produces;
+
+import io.narayana.lra.client.internal.proxy.nonjaxrs.LRAParticipantRegistry;
+
+@Dependent
+public class NarayanaLRAProducers {
+    @Produces
+    public LRAParticipantRegistry lraParticipantRegistry() {
+        return null;
```
[the remark this answers is not in the document]
Comment (mmusgrov) [reply extensions/narayana-lra/runtime/src/main/java/io/quarkus/narayana/lra/runtime/NarayanaLRAProducers.java:12]: Correct

File: integration-tests/narayana-lra/src/test/java/org/acme/quickstart/lra/LRATest.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+                .then()
+                .statusCode(201);
+
+        String body = responseBody.print();
+
+        assertTrue(body.contains("/"), "Response body for a new LRA is not a URL");
+
+        String lra = body.substring(body.lastIndexOf('/') + 1);
+
+        assertNotEquals(0, lra.length(), "new LRA is not a valid URL");
```
[the remark this answers is not in the document]
Comment (mmusgrov) [reply integration-tests/narayana-lra/src/test/java/org/acme/quickstart/lra/LRATest.java:69]: Not really. The definitive test for the validity of the LRA is on line 25:

```java
        assertTrue(lras.contains(lra), "The Narayana LRA coordinator does not know about an active LRA");
```

So if the lra is invalid then the assert will fail.

**Comment by gsmet** [comment]
I'm closing this one. This extension is probably a better fit for the Quarkiverse rather than core. Except if there are very compelling reasons to get them in core.

**Comment by xstefank** [comment]
@gsmet this is an [MP spec](https://github.com/eclipse/microprofile-lra) outside umbrella similar to reactive messaging, streams ops, graphQL, context propagation which are all in the core. The spec will go 1.0.GA after the MP 4.0 working group is established by the mid Nov. Of course, not sure if this is enough to keep it here.

**Comment by mmusgrov** [comment]
@gsmet I would like to reopen this PR but the "Reopen pull request" button is not available, are you able to reopen it for me?

1. I want to use the same PR because the conversation history is important a) for putting my latest updates into context, b) will help the reviewer, and c) will avoid the need to cover the same ground we discussed in the two original PRs (this one linked to issue [4400](https://github.com/quarkusio/quarkus/issues/4400) and its pr [9108](https://github.com/quarkusio/quarkus/pull/9108).
2. I want to target core because compensating transactions are an essential component for building reliable services (and in this respect is similar/complementary to the [narayana-jta](https://github.com/quarkusio/quarkus/tree/main/extensions/narayana-jta) extension)

**Comment by mmusgrov** [comment]
I open pr #17903  for this work

**Pull request closed without merging**